### ISSUE TITLE
pathchk: write unit test to improve coverage

### ISSUE THREAD

**Issue by sylvestre** [uutils/coreutils#1855]
We almost don't have tests for pathchk. We should have more:
https://codecov.io/gh/uutils/coreutils/src/master/src/uu/pathchk/src/pathchk.rs

Should be added here:
https://github.com/uutils/coreutils/blob/master/tests/by-util/test_pathchk.rs

**Comment by Arcterus** [comment]
As an aside, I took a look at the code, and it's rather inefficient.  It seems to allocate a bunch of `String`s when it could just use `&str` (among other similar issues).  It should also use `OsString`/`OsStr` (and thus `clap` instead of `getopts`) to avoid croaking on non-UTF-8 paths.

**Comment by sivachandran** [comment]
@sylvestre I am trying to improve the coverage. I added more tests but grcov doesn't seem to refresh/update the report as per the new tests.

I run grcov with the following commands
```
$ export RUSTC_BOOTSTRAP=1
$ export RUSTFLAGS="-Zinstrument-coverage"
$ cargo test --features feat_os_unix test_pathchk
$ grcov . -s . --binary-path ./target/debug/ -t html --branch --ignore-not-existing --ignore build.rs -o ./target/debug/coverage/
$ open ./target/debug/coverage/index.html
```

I can confirm the new tests are getting executed as I seem. them pass/fail in `cargo test ...` output. I wonder why grcov report is not getting updated and what am I doing wrong. 

I appreciate any help in troubleshooting the issue.

**Comment by sylvestre** [comment]
@sivachandran I guess you have to make cargo clean between runs?

**Comment by sivachandran** [comment]
@sylvestre The report says no source is covered(0%) after `cargo clean`, `cargo build --features feat_os_unix` and `cargo test --features feat_os_unix test_pathchk`. The report lists all files and directories though. Any guess what I'm missing?

**Comment by marco-c** [comment]
@sivachandran I'll try too, but in the meantime can you try gcov-based coverage instead of source-based? https://github.com/mozilla/grcov#example-how-to-generate-gcda-files-for-a-rust-project

**Comment by sivachandran** [comment]
Thanks @marco-c. It worked when I use gcov-based coverage.

**Comment by sivachandran** [comment]
Achieved the below coverage with changes in [PR 1953](https://github.com/uutils/coreutils/pull/1953)
![image](https://user-images.githubusercontent.com/292921/112750010-95e39400-8fe3-11eb-86c6-2f62806b34f4.png)

Strangely, the coverage shows lines of a multi-line function call are not covered despite indicating the first is covered. I'm not sure how can I fix this.
![image](https://user-images.githubusercontent.com/292921/112750119-36d24f00-8fe4-11eb-96d3-4467fd3bcc88.png)

**Comment by sivachandran** [comment]
As the changes are merged(commit dd0addb9c12e23a43818a1bfe14559921da59a20), the issue can be closed.

**Comment by JnRouvignac** [comment]
@sylvestre it looks like this issue is also done now.
